confidential-datahub/docs/SEALED_SECRET.md

```diff
@@ -41,27 +41,35 @@ The format of the KMS type Sealed Secret is
 	"wrap_type": "A256GCM",
 	"iv": "xxx",
 	"annotations": {
-		"region_id": "cn-hangzhou",
-		"endpoint": "xxx",
-		"signature_method": "xxx"
+		"operation_context": {
+			"algorithm": "xxx"
+		},
+		"provider_setting": {
+			"kms_instance_id": "xxx"
+		}
```

Comment:
Ah, yes. At first I want to put all the information about the kms information into the annotation field.

Soon when we did the coding things, we found that there are two different parts of information we need:
Type 1. Parameters to init the KMS client.
Type 2. Parameters of this enc/decryption

The logic is as following:
1. user's client tool side:
a. use Type 1 parameter as the public init parameters of the client
b. the user provides private parameters (like credentials) of the client
c. create the client object
d. use the client to do encryption operation. The encryption will return the extra parameters for the encryption (`operation_context`), like `iv`. (some KMS will generate `iv` server side, and some we need to manually generate in the impl of `encryption` randomly and send to KMS)
e. the client can export its information (`provider_settings`), which will help the cdh to init a client.
f. wrap `operation_context` and `provider_settings` into the annotations and generate the sealed secret

2. in-pod (CDH)
a. when getting a sealed secret, it parses the `operation_context` and `provider_settings` in annotations.
b. use the `provider_settings` as the public init parameters of the client
c. read the hardcoded path to find the private parameters (like credentials) of the client. (Not like the userside where we can directly provide the private parameters directly in the commandline para, in-pod we should read from a safe place. Here we choose the in-memory fs path. The question is how we put these files in that? It can be fetches from KBS. In the [architecture ](https://github.com/confidential-containers/guest-components/blob/main/confidential-datahub/docs/Architecture.svg), we have an `auth layer`. The components in this layer is to get the credentials that is used by ResourceClients like KMS. At first, we can impl KBS to be this one -- KBS client fetch the private parameters from KBS and put them to the specified in-memory fs path)
d. create the client object
e. use the client object to do decryption with the `operation_context` in `annotations`

So, we are actually classifying the fields in annotations for logic decoupling, which is potentially helpful for use cases beyond coco (yes, `kms` it self can be used as a crate). A `serde_json::Map<String, Value>` leaves the extensibility for the two different kinds of parameters.

About the name:
`encryption_context` is more accurate for KMS. But when taking `vault` into consideration, it does not have "encryption". So to make it more generic, the name `operation_context` is chosen to refer to either the context of a KMS enc/decryption or a vault setting/getting operation.

`provider_settings` is the same reason.

Reply:
`operation` is pretty vague. What about calling it `secret_settings` or even just `settings`?

Reply:
How about `secret_setting`. Actually I do not have a good name in my mind either.

Reply:
> These fields are a bit confusing to me. I don't think they were in the original RFC. Will all KMSes need the same annotations? Why separate into `operation_context` and `provider_setting` (should be `provider_settings`?)

Yes, `provider_settings` is needed to initialize an KMS client, we can move to upper level as mandatory, and `annotations` can only save optional data like `secret_settings` or `operattion_context` which can be vendor specific for different KMS implementation.  @Xynnn007

Reply:
> > These fields are a bit confusing to me. I don't think they were in the original RFC. Will all KMSes need the same annotations? Why separate into `operation_context` and `provider_setting` (should be `provider_settings`?)
> 
> Yes, `provider_settings` is needed to initialize an KMS client, we can move to upper level as mandatory, and `annotations` can only save optional data like `secret_settings` or `operattion_context` which can be vendor specific for different KMS implementation. @Xynnn007

Sounds rationale to me. Let me apply this change to the code and plz take another look if it is straightforward and acceptable.

Reply:
force pushed. I think @arronwy gets the core issue. We cannot create a kms client with only `provider` field, but also some other metadata is required. Thus an extra `provider_settings` key-value map is introduced in the sealed secret format. We do not use `secret_settings` anymore, instead we put all other things into the origin optional `annotations` field.